Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift

```diff
@@ -0,0 +1,530 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import Dispatch
+import NIO
+import NIOHTTP1
+
+extension ByteBuffer {
+    init(string: String) {
+        self = ByteBufferAllocator().buffer(capacity: string.utf8.count)
+        self.write(string: string)
+    }
+}
+
+private class MessageEndHandler<Head: Equatable, Body: Equatable>: ChannelInboundHandler {
+    typealias InboundIn = HTTPPart<Head, Body>
+
+    var seenEnd = false
+    var seenBody = false
+    var seenHead = false
+
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head:
+            XCTAssertFalse(self.seenHead)
+            self.seenHead = true
+        case .body:
+            XCTAssertFalse(self.seenBody)
+            self.seenBody = true
+        case .end:
+            XCTAssertFalse(self.seenEnd)
+            self.seenEnd = true
+        }
+    }
+}
+
+/// Tests for the HTTP decoder's handling of message body framing.
+///
+/// Mostly tests assertions in RFC 7230 § 3.3.3.
```

Comment:
@Lukasa nit: consider linking the rfc directly via markdown. (feel free to ignore tho)